docs/connections.md

```diff
@@ -0,0 +1,597 @@
+# General Rules
+This document describes language rules that are relevant to connection discovery
+and operations. The purpose is not to specify the whole language. Just enough
+for IDE team members to be able to reason where identifiers are introduced and
+what entity identifier usage refers to.
+
+This is the base allowing IDE to describe what connections are in the displayed
+graph. The covered topics are mostly identifiers, scopes and their interactions.
+
+
+## Identifier
+Identifier is a name that denotes a value, i.e. that is bound to a value. The
+compiler also keeps track of its type, i.e. the set of possible values. Syntactically we recognize:
+* variable names, being names that do not contain upper-cased characters, like `foo2`
+  or `make_new`;
+* constructor names, which are like variables but with first character and every
+  character directly following underscore upper-cased (e.g. `Foo2` or
+  `Make_New`);
+* operator names, being built solely from operator symbols (e.g. `+` or `<$>`).
+  Specifically, an operator name may contain only characters from the following
+  set: `!$%&*+-/<>?^~|:\,.()[]{}=`. Not every sequence of these characters is a valid operator name, as they can collide with other language constructs.
+  
+Any other names that do not match requirements above, like `HTTP`, `foO` or
+`Make_new` are not allowed and their behavior will not be specified.
+
+In non-pattern contexts, referring to the existing identifier is
+case-insensitive. So `foo` can be referred to either as `foo` or `Foo` with no
+difference.  In pattern context, lower-cased names are used to introduce a
+binding (or a constraint), while upper-cased name will refer to an already bound
+identifier. In short, variable name is allowed to shadow, while constructor 
+always unambiguously refers to an externally visible identifier.
+
+Operator names behave either as:
+* variable names when placed in a prefix position (e.g. `+` in `+ ab`);
+* or as constructor names when placed in an infix position (e.g. `,` in `a,b`).
+
+Number and text literals (like `5` or `"Hello"`) are treated as
+constructor names.
+
+Identifiers can be introduced by a binding (e.g. using assignment or lambda
+argument matching) or when adding type constraints (using type ascriptions). 
+
+## Scope
+Scope is the span in the code which shares the identifiers set of visible,
+available identifiers. Scopes are a span-tree structure, covering the whole program code.
+
+Nested scope is allowed to:
+* access identifiers from the outer scope;
+* shadow identifiers from nested scope with a new binding;
+* introduce new constraints on the identifiers from parent (or own) scopes.
+
+The same identifier may be bound multiple times in the same scope (overloading).
```

Comment:
This is currently not supported outside the global scope. We have an agreement that methods can only be defined in the global scope, so this seems fine to me. If that design decision has changed please let me know.

Reply:
Currently it's a tentative 2. There are technical obstacles given the current way method lookup is performed in the interpreter, but also there would be some typechecking obstacles to doing it which I'm a little unclear on given I'm somewhat divorced from that work right now. The design _currently_ forbids, it, and we'd have to solve these technical challenges to allow it.